Disable inactive IAM users

jsf9k opened the pull request:
## 🗣 Description ##

This pull request adds the [cisagov/disable-inactive-iam-users-tf-module](https://github.com/cisagov/disable-inactive-iam-users-tf-module) Terraform module to periodically disable inactive IAM users.

This module creates an EventBridge event rule that triggers at a fixed cadence, kicking off a Lambda function that disables inactive IAM users.

See also:
- cisagov/provisionaccount-role-tf-module#45
- cisagov/disable-inactive-iam-users-tf-module#1
- cisagov/disable-inactive-iam-users-lambda#1
- cisagov/cool-accounts-userservices#39

## 💭 Motivation and context ##

We have a POAM about disabling inactive users in IAM.  Partially resolves cisagov/cool-system-internal#143.

## 🧪 Testing ##

I applied these changes to all of our COOL static accounts and all of our COOL staging dynamic accounts without incident.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##

- [x] Revert to using the default branch of [cisagov/provisionaccount-role-tf-module](https://github.com/cisagov/provisionaccount-role-tf-module) once cisagov/provisionaccount-role-tf-module#45 has been merged.
- [x] Revert to using the default branch of [cisagov/disable-inactive-iam-users-tf-module](https://github.com/cisagov/disable-inactive-iam-users-tf-module) once cisagov/disable-inactive-iam-users-tf-module#1 has been merged.

## ✅ Post-merge checklist ##

- [x] Apply these changes to all accounts.

jsf9k committed:
Temporarily use a non-default branch of cisagov/provisionaccount-role-tf-module

This is being done for testing purposes and can be reverted once
cisagov/provisionaccount-role-tf-module#45 has been merged.

jsf9k committed:
Add cisagov/disable-inactive-iam-users-tf-module to periodically disable inactive IAM users

This module creates an EventBridge event rule that triggers at a fixed
cadence, kicking off a Lambda function that disables inactive IAM
users.

Note that we are temporarily using a non-default branch of
cisagov/disable-inactive-iam-users-tf-module.  This is being done for
testing purposes and can be reverted once
cisagov/disable-inactive-iam-users-tf-module#1 has been merged.

jsf9k committed:
Update Terraform documentation for these changes

jsf9k committed:
Add a bucket to the master account where Lambda zip files can live

This is intended for Lambdas that are shared across the org, so the
bucket policy allows read access for any account in the org.

jsf9k committed:
Create a role for writing to the Lambda bucket

This role delegates access to the Users account.

jsf9k committed:
Correct bucket policy principal

We want the Terraform to render as "principal": "*", which requires us
to wildcard the type.  See here for more details:
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/iam_policy_document#principals-and-not_principals

jsf9k committed:
Store the Lambda deployment package in a bucket instead of locally

jsf9k committed:
Prefer "Lambda deployment package" to "Lambda zip"

jsf9k committed:
Revert to using the default branch of cisagov/provisionaccount-role-tf-module

We can do this now that cisagov/provisionaccount-role-tf-module#45 has
been merged.

jsf9k committed:
Revert to using the default branch of cisagov/disable-inactive-iam-users-tf-module

We can do this now that cisagov/disable-inactive-iam-users-tf-module#1
has been merged.

jsf9k committed:
Add cisagov/disable-inactive-iam-users module to audit account

jsf9k committed:
Add cisagov/disable-inactive-iam-users module to dns account

jsf9k committed:
Add cisagov/disable-inactive-iam-users module to images account

jsf9k committed:
Add cisagov/disable-inactive-iam-users module to log-archive account

jsf9k committed:
Add cisagov/disable-inactive-iam-users module to master account

jsf9k committed:
Add cisagov/disable-inactive-iam-users module to shared-services account

jsf9k committed:
Add cisagov/disable-inactive-iam-users module to terraform account

jsf9k committed:
Add cisagov/disable-inactive-iam-users module to the dynamic accounts

dav3r commented on master/lambda_bucket.tf:
As discussed late yesterday, all of the Lambda-related resources should reside in the Terraform account, not here.  This is because the Terraform account is the first account that we must bootstrap when we spin up an instantiation of the COOL.

jsf9k replied:
Please see commit [f12d925](https://github.com/cisagov/cool-accounts/pull/147/commits/f12d9250e32c7ea6f7251e47d340a4962def4d4d).

dav3r replied:
Excellent. 👍

dav3r commented on master/variables.tf:
```diff
@@ -63,3 +79,15 @@ variable "tags" {
 }
+
+variable "write_lambda_bucket_role_description" {
+  default     = "Allows sufficient permissions to write to the bucket that contains Lambda deployment packages] in the Master account."
```
Remove errant keystroke here:
```suggestion
  default     = "Allows sufficient permissions to write to the bucket that contains Lambda deployment packages in the Master account."
```

jsf9k replied:
Please see commit [d0b4df4](https://github.com/cisagov/cool-accounts/pull/147/commits/d0b4df45b5f18681a7f012b82a186642bdf10ff7).

dav3r reviewed (changes requested):
Everything looks good, except for the account that the Lambda jazz lives in.

dav3r commented:
I also realized that https://github.com/cisagov/disable-inactive-iam-users-tf-module should be included in https://github.com/cisagov/cool-accounts-userservices as well.

jsf9k committed:
Move Lambda bucket from master account to Terraform account

jsf9k committed:
Add a policy to all accounts that allows reading from the Lambda bucket

Note that the one place where this policy is not required is the
Terraform account, since it already has full access to S3 in order to
create the bucket.

jsf9k committed:
Remove errant character

Co-authored-by: Dave Redmin <david.redmin@gwe.cisa.dhs.gov>

jsf9k committed:
Modernize state bucket configuration

jsf9k commented:
> I also realized that https://github.com/cisagov/disable-inactive-iam-users-tf-module should be included in https://github.com/cisagov/cool-accounts-userservices as well.

Please see cisagov/cool-accounts-userservices#39.

dav3r commented on audit/README.md:
```diff
@@ -95,8 +99,12 @@ future changes by simply running `terraform apply
+| lambda\_key | The S3 key associated with the Lambda function deployment package to disable inactive IAM users. | `string` | n/a | yes |
 | provisionaccount\_role\_description | The description to associate with the IAM role that allows sufficient permissions to provision all AWS resources in the Audit account. | `string` | `"Allows sufficient permissions to provision all AWS resources in the Audit account."` | no |
 | provisionaccount\_role\_name | The name to assign the IAM role that allows sufficient permissions to provision all AWS resources in the Audit account. | `string` | `"ProvisionAccount"` | no |
+| read\_lambda\_bucket\_policy\_description | The description to associate with the IAM role that allows read-only access read-only access to the bucket in the Terraform account containing Lambda deployments. | `string` | `"Allows read-only access read-only access to the bucket in the Terraform account containing Lambda deployments."` | no |
```
Remove duplicated words:
```suggestion
| read\_lambda\_bucket\_policy\_description | The description to associate with the IAM role that allows read-only access to the bucket in the Terraform account containing Lambda deployments. | `string` | `"Allows read-only access to the bucket in the Terraform account containing Lambda deployments."` | no |
```

I'm only flagging it here, you'll have to correct it everywhere.

jsf9k replied:
Please see commit [6e282a6](https://github.com/cisagov/cool-accounts/pull/147/commits/6e282a6b7957fc0dbab85bf4b4959705e35e9c23).

dav3r commented on audit/read_lambda_bucket_policy.tf:
```diff
@@ -0,0 +1,23 @@
+# ------------------------------------------------------------------------------
+# Create the IAM policy that allows all of the permissions necessary
+# to read from the bucket containing the COOL lambda deployments.
```
Capitalization:
```suggestion
# to read from the bucket containing the COOL Lambda deployments.
```

I'm only flagging it here, you'll have to correct it everywhere.

jsf9k replied:
Please see commit [6e282a6](https://github.com/cisagov/cool-accounts/pull/147/commits/6e282a6b7957fc0dbab85bf4b4959705e35e9c23).

dav3r reviewed (approved):
This all LGTM aside from those two issues I noted. 👍

jsf9k committed:
Fix typos

Co-authored-by: David Redmin <david.redmin@gwe.cisa.dhs.gov>
